### PR TITLE
Don't suggest removing necessary casts to arguments passed to params parameters

### LINKED ISSUE #10220: Redundant Cast suggests removing cast to argument passed to params parameter when it shouldn't
**Version Used**: 
VS2015 Update 1

(I expect this is a duplicate, but I didn't find it.)

**Steps to Reproduce**:
1. Place this code in a C# Console program:

```
    class Program
    {
        static void Main(string[] args)
        {
            object[] arr = { 1, 2, "3", null, 5 };
            testParams(arr);
            testParams((object)arr);

            if (Debugger.IsAttached)
                Console.ReadLine();
        }

        static void testParams(params object[] ps)
        {
            Console.WriteLine(ps.Length);
        }

    }

```

**Expected Behavior**:
WIth the "redundant" `(object)` cast the output is 5, 1.

**Actual Behavior**:
If you allow the cast to be removed you obviously get 5, 5.

### PR DESCRIPTION
**Pull request by DustinCampbell**
Fixes #10220

cc @dotnet/roslyn-ide

### CODE REVIEW

**Comment by DustinCampbell** [comment]
Anyone from @dotnet/roslyn-ide want to take a look?

File: src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs [hunk 1]
```diff
@@ -145,7 +145,7 @@ private static bool UserDefinedConversionIsAllowed(ExpressionSyntax expression,
             SemanticModel semanticModel,
             CancellationToken cancellationToken)
         {
-            // Special case: When a null literal is cast and passed as the single argument to a params parameter,
+            // Special case: When a cast and passed as the single argument to a params parameter,
```
Comment (rchande) [inline src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs]: You a word

**Comment by rchande** [comment]
:+1:

File: src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs [hunk 2]
```diff
@@ -162,43 +162,40 @@ private static bool UserDefinedConversionIsAllowed(ExpressionSyntax expression,
...
+                }
 
-                    var parameter = argument.DetermineParameter(semanticModel, cancellationToken: cancellationToken);
-                    if (parameter != null && parameter.IsParams)
-                    {
-                        Debug.Assert(parameter.Type is IArrayTypeSymbol);
+                var parameter = argument.DetermineParameter(semanticModel, cancellationToken: cancellationToken);
+                if (parameter != null && parameter.IsParams)
+                {
+                    Debug.Assert(parameter.Type is IArrayTypeSymbol);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs]: I don't see why that has to be the case.  What about in error code if you have:  ```Foo(params int i)```.  Does the compiler guarantee you'll have an array type?

Comment (DustinCampbell) [inline src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs]: I have no idea. I just left the rest of the code the same. I'll take a closer look. Thanks!

File: src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs (hunk 1 above)
Comment (DustinCampbell) [inline src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs]: Thanks!

**Comment by DustinCampbell** [comment]
retest this please

**Comment by agocke** [comment]
@DustinCampbell  Does your argument indexing code work for the following example?

```csharp
    class Program
    {
        static void Main(string[] args)
        {
            M(xs: (int)3, y: 2, x: 1);
        }

        private static void M(int x = 0, int y = 1, params int[] xs)
        {
        }
    }
```

**Comment by DustinCampbell** [comment]
@agocke: Probably not. I didn't change that code dramatically though. It looks like that'd fail right here. https://github.com/dotnet/roslyn/pull/10702/files#diff-6f8f267309a989ee1a5a92a703afe42bR174

**Comment by agocke** [comment]
@DustinCampbell Is that something you would want to fix? :)

**Comment by DustinCampbell** [comment]
To make you happy? Not really? :smile:

**Comment by agocke** [comment]
Just here to make your life more difficult :100:

**Comment by Pilchie** [comment]
:+1:, though this now has conflicts that need resolving.

(Sorry for the delay in reviewing :()

**Comment by DustinCampbell** [comment]
I was actually going to hold off on this until I address @agocke's points.

**Pull request closed without merging**